content/en/blog/_posts/2019-03-27-kube-proxy-subtleties-debugging-an-intermittent-connection-resets.md

```diff
@@ -0,0 +1,169 @@
+---
+title: kube-proxy subtleties: debugging an intermittent connection resets
+date: 2019-03-27
+---
+
+**Author:** Yongkun Gui (Google)
+
+I recently came across a bug that causes intermittent connection resets.  After
+some digging, I found it was caused by a subtle combination of several different
+network subsystems. It helps me understand Kubernetes networking better, and I
+think it’s worthwhile sharing to a wider audience who are interested in the same
+topic.
+
+## The symptom
+
+We saw a user report seeing connection resets using a ClusterIP service to serve
+large files to pods running in their cluster. Initial debugging of the cluster
+did not yield anything interesting: network connectivity was fine and
+downloading the files did not hit any issues. However, when we ran the workload
+in parallel across many clients, we were able to reproduce the problem. Adding
+to the mystery was the fact that when the problem could not be reproduced when
+the workload was run using VMs without Kubernetes. There is a simple app to
+reproduce the issue [here](https://github.com/tcarmet/k8s-connection-reset). The
+problem clearly has something to do with Kubernetes networking, but what?
+
+## Kubernetes networking basics
+
+Before digging into this problem, let’s talk a little bit about some basics of
+Kubernetes networking first. When a pod initializes a connection to either
+another pod, a in-cluster service or an external service, Kubernetes handles it
+differently.
```

Comment:
i would personally reformat it as 

> Before digging into the problem, let’s talk a bit about basics of Kubernetes networking as the connection to either another pod, an in-cluster service or an external service is being handled differently by Kubernetes while the pod is being initialized.

Reply:
It's not exactly what I want to say, but I get your suggestion.